twoliter/embedded/img2img

```diff
@@ -1,3 +1,318 @@
+    ;;
+  esac
+
+  # 1 MiB contains 2048 512-byte sectors.
+  sgdisk_start_mib="$((sgdisk_start / 2048))"
+
+  if [[ "${partoff["${part}"]}" -ne "${sgdisk_start_mib}" ]]; then
+    echo "start mismatch between partition plan and disk: '${part}'" >&2
+    exit 1
+  fi
+
+  if [[ "${partsize["${part}"]}" -ne "${sgdisk_size%.0}" ]]; then
+    echo "size mismatch between partition plan and disk: '${part}'" >&2
+    exit 1
+  fi
+done
+
+###############################################################################
+# Section 2: extract needed partitions from the OS image
+
+# Extract the root image from the OS image.
+dd if="${OS_IMAGE}" of="${ROOT_IMAGE}" \
+  count="${partsize["ROOT-A"]}" bs=1M skip="${partoff["ROOT-A"]}"
+
+# Extract the boot partition from the OS image, and dump the contents.
+dd if="${OS_IMAGE}" of="${BOOT_IMAGE}" \
+  count="${partsize["BOOT-A"]}" bs=1M skip="${partoff["BOOT-A"]}"
+debugfs -R "rdump / ${BOOT_MOUNT}" "${BOOT_IMAGE}"
+
+# Extract the EFI partition from the OS image and dump the contents, if needed.
+if [[ "${UEFI_SECURE_BOOT}" == "yes" ]]; then
+  dd if="${OS_IMAGE}" of="${EFI_IMAGE}" \
+    count="${partsize["EFI-A"]}" bs=1M skip="${partoff["EFI-A"]}"
+  mkdir -p "${EFI_MOUNT}/EFI/BOOT"
+  mcopy -i "${EFI_IMAGE}" -sv ::/EFI/BOOT "${EFI_MOUNT}/EFI/"
+  # Remove the original signing certificates to avoid confusion.
+  rm -f "${EFI_MOUNT}/EFI/BOOT/db.cer" "${EFI_MOUNT}/EFI/BOOT/db.crt"
+fi
+
+###############################################################################
+# Section 3: replace needed artifacts
+
+# TODO - CA bundle code goes here
+
+###############################################################################
+# Section 4: update root partition and root verity
+
+# shellcheck disable=SC2312  # mapfile is validated elsewhere
+mapfile -t new_root_artifacts <<<"$(find "${ROOT_MOUNT}" -type f)"
+
+# TODO - Error if no artifacts since there will always at least the CA bundle.
+if [[ -z "${new_root_artifacts[0]}" ]]; then
+  echo "Nothing to write, skipping..."
+else
+  # Write files from the root mount to the root image.
+  for artifact in "${new_root_artifacts[@]}"; do
+    cat <<EOF | debugfs -w -f - "${ROOT_IMAGE}"
+rm ${artifact#"${ROOT_MOUNT}"}
+write ${artifact} ${artifact#"${ROOT_MOUNT}"}
+ea_set ${artifact#"${ROOT_MOUNT}"} security.selinux system_u:object_r:os_t:s0
+EOF
+  done
+fi
+
+# Validate and write root image back to the OS image.
+check_image_size "${ROOT_IMAGE}" "${partsize["ROOT-A"]}"
+dd if="${ROOT_IMAGE}" of="${OS_IMAGE}" \
+  conv=notrunc bs=1M seek="${partoff["ROOT-A"]}"
+
+# Generate a new root verity.
+declare -a DM_VERITY_ROOT
+generate_verity_root "${ROOT_IMAGE}" "${VERITY_IMAGE}" \
+  "${partsize["HASH-A"]}" DM_VERITY_ROOT
+
+# Validate and write root verity back to the OS image. The image size check
+# isn't needed here as it's already done in `generate_verity_root`.
+dd if="${VERITY_IMAGE}" of="${OS_IMAGE}" \
+  conv=notrunc bs=1M seek="${partoff["HASH-A"]}"
```

Comment:
The comment refers to validation but I just see the write?

Reply:
It's on [L316](https://github.com/bottlerocket-os/twoliter/pull/231/files#diff-0b8c4b6920fcc7cd541e1b6cac7893cf86c9b2e80859c9444ebdf3339b5f3196R316) of `imghelper` since it's checked both in `img2img` and `rpm2img`.